include/cell_actor.h

```diff
@@ -0,0 +1,193 @@
+#ifndef POKEPLATINUM_CELL_ACTOR_H
+#define POKEPLATINUM_CELL_ACTOR_H
+
+#include <nitro/gx.h>
+#include <nnsys.h>
+#include <nitro/fx/fx.h>
+
+#include "constants/heap.h"
+
+#define CELL_ACTOR_FLIP_NONE        0
+#define CELL_ACTOR_FLIP_H           1
+#define CELL_ACTOR_FLIP_V           2
+
+#define CELL_ACTOR_ANIM_DATA_SIZE   29
+#define MAX_SIMULTANEOUS_SPRITES    128
+
+enum CellAnimType {
+    CELL_ANIM_TYPE_NONE = 0,
+    CELL_ANIM_TYPE_CELL,
+    CELL_ANIM_TYPE_MULTI_CELL,
+    CELL_ANIM_TYPE_VRAM_CELL,
+};
+
+enum AffineOverwriteMode {
+    AFFINE_OVERWRITE_MODE_NONE = 0,
+    AFFINE_OVERWRITE_MODE_NORMAL,
+    AFFINE_OVERWRITE_MODE_DOUBLE, // Extends the cell actor's drawable area by 2x, use when the cell actor is scaled up
+};
+
+typedef struct CellActorCollection CellActorCollection;
+
+typedef struct CellAnimationData {
+    const NNSG2dCellDataBank *cellBank;
+    const NNSG2dCellAnimBankData *animBank;
+    NNSG2dCellAnimation anim;
+} CellAnimationData;
+
+typedef struct VRamCellAnimationData {
+    NNSG2dCellDataBank *cellBank;
+    const NNSG2dCellAnimBankData *animBank;
+    NNSG2dCellAnimation anim;
+    u32 transferHandle;
+} VRamCellAnimationData;
+
+typedef struct MultiCellAnimationData {
+    const NNSG2dCellDataBank *individualCellBank;
+    const NNSG2dCellAnimBankData *individualAnimBank;
+    NNSG2dMultiCellAnimation anim;
+    const NNSG2dMultiCellDataBank *cellBank;
+    const NNSG2dMultiCellAnimBankData *animBank;
+    NNSG2dNode *nodes;
+    NNSG2dCellAnimation *cellAnims;
+} MultiCellAnimationData;
+
+typedef struct CellActor {
+    VecFx32 position;
+    VecFx32 affineTranslation;
+    VecFx32 affineScale;
+    u16 affineZRotation;
+    u8 affineOverwriteMode;
+    u8 flip;
+    u8 overwriteFlags; // Specifies which of the 'explicit' fields are used. Overwrites data provided by the OAM.
+    u8 explicitPalette; // An explicit palette index.
+    u8 explicitPaletteOffset; // An explicit palette index offset added onto the index specified by the OAM.
+    BOOL explicitMosaic;
+    GXOamMode explicitOamMode;
+    u8 draw;
+    u8 animate;
+    fx32 animSpeed;
+    CellActorCollection *collection; // The collection this actor belongs to
+
+    // This field is supposed to be a union between CellAnimationData, VRamCellAnimationData, and MultiCellAnimationData
+    // but it's actually too small to hold the largest of these types. This should really be u32 animData[31].
+    u32 animData[CELL_ACTOR_ANIM_DATA_SIZE];
+    NNSG2dImageProxy imageProxy;
+    NNSG2dImagePaletteProxy paletteProxy;
+    u32 type;
+    u16 activeAnimID;
+    u8 explicitPriority;
+    u16 priority;
+    NNS_G2D_VRAM_TYPE vramType;
+    struct CellActor *prev;
+    struct CellActor *next;
```

Comment:
**question:** Is this just a doubly-linked list? That also has a `head` pointer? Kind of odd.